### ISSUE TITLE
`kfp-v2` bundle tests fail during recurring run

### ISSUE THREAD

**Issue by orfeas-k** [canonical/kfp-operators#602]
### Bug Description

Intermittently with #601, tests fail during test_create_and_monitor_recurring_run:

```
AssertionError: assert 'FAILED' == 'SUCCEEDED'
  
  - SUCCEEDED
  + FAILED
```

Looking at the juju status, it looks like node could be potentially out of space

```
Model     Controller                Cloud/Region        Version  SLA          Timestamp
kubeflow  github-pr-38fa2-microk8s  microk8s/localhost  3.4.6    unsupported  14:34:1[7](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:8)Z

App                      Version                  Status   Scale  Charm                    Channel       Rev  Address         Exposed  Message
argo-controller                                   active       1  argo-controller          latest/edge   596  10.152.1[8](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:9)3.140  no       
envoy                                             active       1  envoy                    latest/edge   307  10.152.183.161  no       
istio-ingressgateway                              active       1  istio-gateway            latest/edge  1287  10.152.183.151  no       
istio-pilot                                       active       1  istio-pilot              latest/edge  1235  10.152.183.45   no       
kfp-api                                           active       1  kfp-api                                  0  10.152.183.241  no       
kfp-db                   8.0.37-0ubuntu0.22.04.3  active       1  mysql-k8s                8.0/stable    180  10.152.183.8[9](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:10)   no       
kfp-metadata-writer                               active       1  kfp-metadata-writer                      0  [10](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:11).152.183.114  no       
kfp-persistence                                   active       1  kfp-persistence                          0  10.152.183.247  no       
kfp-profile-controller                            active       1  kfp-profile-controller                   0  10.152.183.139  no       
kfp-schedwf                                       active       1  kfp-schedwf                              0  10.152.183.214  no       
kfp-ui                                            active       1  kfp-ui                                   0  10.152.183.248  no       
kfp-viewer                                        active       1  kfp-viewer                               0  10.152.183.170  no       
kfp-viz                                           active       1  kfp-viz                                  0  10.152.183.20   no       
kubeflow-profiles                                 active       1  kubeflow-profiles        latest/edge   456  10.152.183.58   no       
kubeflow-roles                                    active       1  kubeflow-roles           latest/edge   264  10.152.183.57   no       
metacontroller-operator                           active       1  metacontroller-operator  latest/edge   349  10.152.183.185  no       
minio                    res:oci-image@220b31a    blocked    0/1  minio                    latest/edge   376  10.152.183.226  no       0/1 nodes are available: 1 node(s) had untolerated taint {node.kubernetes.io/disk-pressure: }. preemption: 0/1 nodes ...
mlmd                                              active       1  mlmd                     latest/edge   243  10.152.183.225  no       

Unit                        Workload  Agent  Address      Ports          Message
argo-controller/0*          active    idle   10.1.2[11](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:12).74                 
envoy/0*                    active    idle   10.1.211.75                 
istio-ingressgateway/0*     active    idle   10.1.211.76                 
istio-pilot/0*              active    idle   10.1.211.77                 
kfp-api/0*                  active    idle   10.1.211.78                 
kfp-db/0*                   active    idle   10.1.211.82                 Primary
kfp-metadata-writer/0*      active    idle   10.1.211.81                 
kfp-persistence/0*          active    idle   10.1.211.84                 
kfp-profile-controller/0*   active    idle   10.1.211.85                 
kfp-schedwf/0*              active    idle   10.1.[21](https://github.com/canonical/kfp-operators/actions/runs/11838249242/job/32986966903?pr=583#step:7:22)1.86                 
kfp-ui/0*                   active    idle   10.1.211.88                 
kfp-viewer/0*               active    idle   10.1.211.89                 
kfp-viz/0*                  active    idle   10.1.211.90                 
kubeflow-profiles/0*        active    idle   10.1.211.93                 
kubeflow-roles/0*           active    idle   10.1.211.91                 
metacontroller-operator/0*  active    idle   10.1.211.92                 
minio/0*                    unknown   lost   10.1.211.97  9000-9001/TCP  agent lost, see 'juju show-status-log minio/0'
mlmd/0*                     active    idle   10.1.211.95
```

### To Reproduce

Rerun CI from PR https://github.com/canonical/kfp-operators/pull/583

### Environment

Juju 3.4.6
Microk8s 1.29

### Relevant Log Output

```shell
____________________ test_create_and_monitor_recurring_run _____________________
Traceback (most recent call last):
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/runner.py", line 341, in from_call
    result: TResult | None = func()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/runner.py", line 242, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/threadexception.py", line 92, in pytest_runtest_call
    yield from thread_exception_runtest_hook()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/threadexception.py", line 68, in thread_exception_runtest_hook
    yield
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/unraisableexception.py", line 95, in pytest_runtest_call
    yield from unraisable_exception_runtest_hook()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/unraisableexception.py", line 70, in unraisable_exception_runtest_hook
    yield
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/logging.py", line 848, in pytest_runtest_call
    yield from self._runtest_for(item, "call")
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/logging.py", line 831, in _runtest_for
    yield
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/capture.py", line 879, in pytest_runtest_call
    return (yield)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/skipping.py", line 257, in pytest_runtest_call
    return (yield)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/runner.py", line 174, in pytest_runtest_call
    item.runtest()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/python.py", line 1627, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 182, in _multicall
    return outcome.get_result()
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_result.py", line 100, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/_pytest/python.py", line 159, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/home/runner/work/kfp-operators/kfp-operators/.tox/bundle-integration-v2/lib/python3.8/site-packages/pytest_asyncio/plugin.py", line 529, in inner
    _loop.run_until_complete(task)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/runner/work/kfp-operators/kfp-operators/tests/integration/test_kfp_functional_v2.py", line 204, in test_create_and_monitor_recurring_run
    assert monitor_response.state == "SUCCEEDED"
AssertionError: assert 'FAILED' == 'SUCCEEDED'
  
  - SUCCEEDED
  + FAILED
```

**Comment by orfeas-k** [comment]
Adding a `df -h` before every test [showed](https://github.com/canonical/kfp-operators/actions/runs/11909109360/job/33193929888#step:6:190) that this is a disk size issue:
```
# df -h before test_create_and_monitor_recurring_run
Filesystem      Size  Used Avail Use% Mounted on
/dev/root        73G   72G  1.4G  99% /
devtmpfs        7.9G     0  7.9G   0% /dev
tmpfs           7.9G  4.0K  7.9G   1% /dev/shm
tmpfs           1.6G  3.2M  1.6G   1% /run
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           7.9G     0  7.9G   0% /sys/fs/cgroup
/dev/sdb15      105M  6.1M   99M   6% /boot/efi
/dev/loop0       64M   64M     0 100% /snap/core20/2379
/dev/loop1       39M   39M     0 100% /snap/snapd/21759
/dev/loop2       92M   92M     0 100% /snap/lxd/29619
/dev/sda1        74G   28K   70G   1% /mnt
tmpfs           1.6G  4.0K  1.6G   1% /run/user/1001
/dev/loop3      105M  105M     0 100% /snap/core/17[200](https://github.com/canonical/kfp-operators/actions/runs/11909109360/job/33193929888#step:6:201)
/dev/loop4       74M   74M     0 100% /snap/core22/1663
/dev/loop5      105M  105M     0 100% /snap/lxd/30130
tmpfs           1.0M     0  1.0M   0% /var/snap/lxd/common/ns
/dev/loop6       95M   95M     0 100% /snap/juju/28491
/dev/loop7       28M   28M     0 100% /snap/charm/712
/dev/loop8       59M   59M     0 100% /snap/charmcraft/4914
/dev/loop9      256K  256K     0 100% /snap/jq/6
/dev/loop10     1.5M  1.5M     0 100% /snap/juju-bundle/25
/dev/loop11      13M   13M     0 100% /snap/juju-crashdump/271
/dev/loop12     163M  163M     0 100% /snap/microk8s/7396
/dev/loop13      13M   13M     0 100% /snap/kubectl/3446
```

This explains also why `minio` unit is on blocked after tests complete
```
minio/0*                    blocked      idle                   9000-9001/TCP  0/1 nodes are available: 1 node(s) had untolerated taint {node.kubernetes.io/disk-pressure: }. preemption: 0/1 nodes ..
```
with its pod being scheduled and the node having a `NoSchedule` taint due to `disk-pressure`.
```
apiVersion: v1
kind: Node
metadata:
  annotations:
    node.alpha.kubernetes.io/ttl: "0"
    projectcalico.org/IPv4Address: 10.134.241.1/24
    projectcalico.org/IPv4VXLANTunnelAddr: 10.1.84.64
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: "2024-11-20T07:48:12Z"
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/os: linux
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: fv-az1423-406
    kubernetes.io/os: linux
    microk8s.io/cluster: "true"
    node.kubernetes.io/microk8s-controlplane: microk8s-controlplane
  name: fv-az1423-406
  resourceVersion: "10757"
  uid: b2941984-057f-4273-9150-698e1a075152
spec:
  taints:
  - effect: NoSchedule
    key: node.kubernetes.io/disk-pressure
    timeAdded: "2024-11-20T08:48:26Z"
...
```

We plan to refactor our tests soon and build each charm in a separate runner which will eliminate the issue of the runner being out of space. In the meantime, we will workaround this by deleting the lxd instances after build and deployment has been completed. This will be achieved with a snippet based on https://discourse.charmhub.io/t/how-to-quickly-clean-unused-lxd-instances-from-charmcraft-pack/15975
```
    lxc_instances = sh.lxc.list(project="charmcraft", format="json")
    lxc_instances_charmcraft = jq.compile('.[] | select(.name | startswith("charmcraft-")) | .name').input_text(lxc_instances).all()
    for instance in lxc_instances_charmcraft:
        print(f"Deleting {instance}")
        sh.lxc.delete(instance, project="charmcraft")
``` 

In order to avoid this behaviour being the default one, we will also introduce a flag that will enable/disable this.

**Comment by orfeas-k** [comment]
Closed by #616
